swandaskcluster/cluster.py

```diff
@@ -0,0 +1,61 @@
+
+from . import config
+from .utils import SchedulerConfig
+
+from dask_lxplus import CernCluster
+
+import dask
+
+
+class SwanHTCondorCluster(CernCluster):
+    '''
+    Class that configures an HTCondorCluster to be used from SWAN.
+    '''
+
+    config_name = 'swan'
+
+    def __init__(self,
+                 worker_image = None,
+                 job_extra = {},
+                 scheduler_options = {},
+                 **base_class_kwargs):
+        '''
+        Configures the Dask workers to be used by SWAN clients and dynamically
+        obtains the necessary addresses for a Dask scheduler that runs in a
+        SWAN user session.
+        '''
+
+        # Worker configuration
+        worker_image = worker_image or \
+                       dask.config.get(
+                           f'jobqueue.{self.config_name}.worker-image')
+
+        config_job_extra = dask.config.get(
+                               f'jobqueue.{self.config_name}.job-extra')
+
+        # TODO: set a value for 'log_directory'
+
+        # Scheduler configuration
+        self._scheduler_config = SchedulerConfig()
+        private_hostname = self._scheduler_config.get_private_hostname()
+        contact_hostname = self._scheduler_config.get_contact_hostname()
+
+        try:
+            port = self._scheduler_config.get_port()
+
+            scheduler_options['host']            = f'{private_hostname}:{port}'
+            scheduler_options['contact_address'] = f'{contact_hostname}:{port}'
+
+            super().__init__(worker_image = worker_image,
+                             job_extra = { **config_job_extra, **job_extra },
+                             scheduler_options = scheduler_options,
+                             **base_class_kwargs)
+        except Exception:
+            # Something went wrong.
+            # Release the port so that it can be given to other processes and
+            # re-raise
+            self._scheduler_config.release_port()
```

Comment:
If the exception is in requesting a port (we might've used all of them), is the error message clear enough for the users? Also, is the message propagated to the UI?

Reply:
yeah maybe it's worth distinguishing between exceptions and printing something informative in case of a `NoPortsException` ?

Reply:
Ok I'll make several exception cases, anyway if the cluster is created from the Dask lab extension, the user might not even see the error (to be tested).

Reply:
The more fine-grained exception handling is there, I tested it when creating the cluster in a notebook manually but I couldn't test it yet from the UI because we still have some issue to iron out when creating clusters. If the Dask lab extension shows exceptions thrown when creating a cluster, it will be visible to users. If not, at least the exception message should appear in the container logs (hopefully).

Reply:
Please let me know if you have further comments about the PR.